### ISSUE TITLE
Ensure Docker isn't relying on GitHub Services

### ISSUE THREAD

**Issue by bmw** [certbot/certbot#6630]
I got this email from Docker over the holidays:

> Docker Hub Automated Builds affected by Github Services deprecation
We have identified that your Docker Hub account contains Classic Automated Builds that are using GitHub Services.
>
> GitHub Services are being deprecated on January 31st, 2019 and will experience an interruption of service from January 7th to January 13th, 2019, inclusive. Our attempt to automatically upgrade your Automated Builds to use GitHub Webhooks has failed due to an issue accessing your GitHub repositories.
>
> We advise you to take action to reestablish your build web hooks before Monday, January 7th, 2019 to avoid affecting your workflow.
>
> Thank you for your attention in this matter. If you have any questions or concerns please contact us at hub@docker.com.

It looks like our Docker setup is relying on GitHub services. If I remember correctly, I did things these way to not give Docker a ton of write permissions on our repo. If giving them these permissions is the only way to do things now, we could consider it, but I think it should be avoided if possible.

**Comment by adferrand** [comment]
I had a discussion recently with @bmw.

I think the option 2 is the best so far. With this, you can invert the logic and push a docker image from the CI into Docker Hub. However with this, description from README.md and Dockerfile in Docker Hub are not updated anymore after a build. And no API is settled to do that except during an Automated Build.

I think that description does not need to match README.md from Github, as it is not the same purpose, as the description should focus on how to use the Docker, when README.md is about using the project (including Docker) + how to contribute.

The absence of Dockerfile is easily compensated by the visual link to Github project. I do not bother personally, but not everyone agree on that matter.

In general, it is true, in particular in corporate environments, that only official Dockers are trusted. You do not run your business on some unknown guy that could stop maintaining because he does not have time anymore ^^ And security issues in containers become a increasing critical issue.

So making the effort to apply to official Docker images, could really improve the acceptance of dockerized Certbot instances.

**Comment by bmw** [comment]
I think I prefer option 2 for now. I'm not a fan of option 1 either. I could be convinced to go with option 3 if we think setting up a simple yet maintainable version of it is easier than setting up the same for option 2, but I think option 2 is the best approach out of those three in the long term.

Another potential option might be to send remote build triggers to Docker Hub when we want them to rebuild the images. There's a brief section on this at https://docs.docker.com/v17.12/docker-hub/builds/. This could be done in a simple script either in CI or as part of the release process or we could maybe set it up to run in GitHub. This may be relying on the GitHub Services integration though. I'm not sure.

Switching to official Docker images may be worth considering in the long term, but I personally don't think it's worth investing the effort in doing that now. I think we have other, higher impact things we can work on.

Getting rid of our Docker images entirely is potentially an option. The main reason we have them to begin with is because they are so easy to offer. For much of what Certbot does, running in Docker doesn't make sense. If Docker images are becoming more difficult to offer, we could consider deprecating them entirely. As you can see from my stats at https://github.com/certbot/certbot/issues/6504#issuecomment-451610741, they certainly aren't used very much compared to the other installation approaches we have.

The one area where I think it'd be temporarily painful to get rid of our Docker images now is DNS plugins. Currently our instructions at https://certbot.eff.org say that if packages for our DNS plugins aren't available for your system, you can use Docker. If we get rid of Docker, we're removing the one option those users have to get these plugins right now. At the same time though, since so few people are using these plugins (see https://github.com/certbot/certbot/issues/6504#issue-380825681), maybe this is OK.

I'm curious what other people think.

**Comment by joohoi** [comment]
> Switching to official Docker images may be worth considering in the long term, but I personally don't think it's worth investing the effort in doing that now. I think we have other, higher impact things we can work on.

I do agree, the option 2) looks like the best (at least) short term solution. I'm not in favor of getting rid of the Docker images completely.

**Comment by adferrand** [comment]
I confirm. Full read+write access to GitHub account is mandatory for Docker Hub Automated Builds.

**Comment by bmw** [comment]
That's so unfortunate.

@joohoi, do you have any major concerns with this approach?

**Comment by joohoi** [comment]
No, I don't, and even though I didn't communicate that clearly enough in the original comment, this was what I came up as well. The automated builds are naturally the only way to publicize the Dockerfile in the DockerHub altogether. 

Because of this, I feel that setting up the CI pipeline to update Docker images is the best short term solution, but in the long term we should re-evaluate the possibility of getting a official Docker image - maily because of the stated trust implications.

**Comment by bmw** [comment]
I can think of two minor problems:

1. Are we comfortable giving Travis the keys to push to our Docker Hub repository?
2. How do we ensure Travis runs on the release tags so that it can build the stable versions of the Docker images? With our current release process, this is guaranteed to happen for non-bug fix releases because the release tag is the start of a new `0.<minor version>.x` branch which runs tests but not bug fix releases. See https://github.com/certbot/certbot/pull/6632 and https://github.com/certbot/certbot/pull/6564 respectively. Is there an easy way to fix this on Travis' end or would we need to modify our release process?

Another potential (temporary) solution that avoids both of these problems would be to write a script we can run as part of the release process to build and push the Docker images to Docker Hub.

What do you all think?

**Comment by joohoi** [comment]
Good points. Looking more into it;

> Are we comfortable giving Travis the keys to push to our Docker Hub repository?

No, I'm not much of a fan of this.

The way to handle this would be to use Travis CLI, which - in my opinion - seems to be just added complexity with questionable benefit. This would also be a largely manual operation to handle within the release process. I think we'd be better off after all by scripting manual step around it for the release process instead of using Travis to push the images there. The tests we currently have must be green for the release process anyways.

> I suspect this to be an unpopular option, but another idea here is to stop using Docker Hub. We still have a very old setup on quay.io which is still regularly updating using webhooks rather than GitHub services.

Yeah, I think having Certbot images in DockerHub is beneficial, and AFAIK that's basically the only commonly used portal for Docker image discovery. To be honest, the last I heard about quay.io was two years ago, in some of our internal communcation.

**Comment by bmw** [comment]
@adferrand, do you have any major concerns about us writing a script we run locally as part of the release process or after it to update these images?

**Comment by joohoi** [comment]
We could also add a manual step, triggering the buld from DockerHub site manually.

**Comment by adferrand** [comment]
Sadly this @joohoi, still needs Docker Hub to be connected, with write access, to our GitHub repositories. Without that, you do not have access to the webhook that triggers a build.

**Comment by adferrand** [comment]
Also about Quay.io, yes, it is not very visible compared to Docker Hub. I do not know also their policy about GitHub accesses to execute and publish new dockers.

**Comment by adferrand** [comment]
For your proposition @bmw, yes, I think it the best solution. Anyway, pushing the Docker during the CI process is not the safest thing to do, as only a part of the tests are executed continuously. So it is likely that a broken Certbot version would be pushed to the `latest` tag. Many users use directly the `latest` tag (on which I actively militate against), so will retrieve also the broken Certbot version.

In a way or another, the Dockers published need to be verified versions, with `latest` been the latest published verified version. So all of this basically happens during the release process, and I do not see specific interest to have this during the CI, so on Travis CI or AppVeyor.

To sum up, no major concern, you should indeed just put an extra step after PyPi release for instance, to build and push the Dockers with appropriate tags, to Docker Hub. And the credentials are not shared between the services, just set on you machine locally during the release process.

As a perspective, the release script will ultimately be translated in Python one day. Also all manual steps should eventually go in a Continous Delivery approach, by been heavily automated. At the end, a CD pipeline could be implemented in a self hosted service, like Jenkins, directly by the EFF servers, to avoid to store sensible credentials in foreign servers.

**Comment by joohoi** [comment]
> Sadly this @joohoi, still needs Docker Hub to be connected, with write access, to our GitHub repositories. Without that, you do not have access to the webhook that triggers a build.

It is possible if you trigger the build from the web UI of Docker Hub manually. This is what I do for `acme-dns`. The Dockerfile is visible as well: https://hub.docker.com/r/joohoi/acme-dns/dockerfile

**Comment by adferrand** [comment]
But your docker repo is connected to you github repo, so it has write access on it, otherwise you would not have a link to github in the Docker project.

**Comment by joohoi** [comment]
It was initially connected, but there are no active OAuth credentials in the project. Docker Hub UI doesn't list it as connected, and the GitHub keys have been revoked from its side.

**Comment by adferrand** [comment]
Tricky, I did not test that! In this case, yes clearly we could try that. I will test if the webhook still works.

**Comment by adferrand** [comment]
I just revoke the DockerHub App from the Authorized OAuth Apps in my personal GitHub account.

On DockerHub side, the webhook to trigger a build on a specific tag is still working, I successfully triggered a build using the curl command provided by DockerHub as an example.

**Comment by bmw** [comment]
@adferrand, just to make sure, you're referring to the build triggers described at https://docs.docker.com/v17.12/docker-hub/builds/#remote-build-triggers right?

**Comment by adferrand** [comment]
Yes, I am referring to that.

**Comment by bmw** [comment]
Just successfully tested this on Certbot's Docker Hub repo to update the images to reflect the new release.

One concern I have (which existed before so maybe we don't need to solve it now) is making sure the DNS plugin images are built using the updated Certbot image. When we do a release, we want to trigger builds for both Certbot and the DNS plugin images. The DNS plugin images are based on the Certbot image and the plugins may not even work with older versions of Certbot. How do we ensure the DNS plugin image gets built using updated Certbot image?

I'll think more about this, but if anyone has ideas, please share!

**Comment by adferrand** [comment]
One way is to do multistage builds: https://www.google.fr/url?sa=t&source=web&rct=j&url=https://docs.docker.com/develop/develop-images/multistage-build/&ved=2ahUKEwjlxf7u86rgAhWqzIUKHbqpAxoQFjAAegQIBBAB&usg=AOvVaw2hJoNXmhILBohxghui71Fv

You build a first Docker using controlled version of some sources (so here the checkout of specific version of Certbot), then you use this Docker content to build another Docker, that is the Docker you publish.

This is done in one unique Dockerfile.

**Comment by adferrand** [comment]
Otherwise you can have a CI/CD pipeline, with a first stage that build docker for certbot and publish for a version tag. Then multiple stages in parallel for each dns plugin, that use for their FROM directive, the certbot docker you just published. And then you publish them also.

As you can inject environment variables everywhere in a Dockerfile, sharing the version between the stages of the pipeline will allow you to coordinate the publication on a specific version, while keeping the dockerfile agnostic to a given version.

**Comment by bmw** [comment]
> This is done in one unique Dockerfile.

At least the way I'm thinking about it, I think this might be kind of hard/awkward with our current setup with many DNS plugins each in their own Dockerfile.

I think something like what you describe in your 2nd post might work though. What if we write a script that:

1. Determines the latest tag on the Certbot repo.
2. Checks if a Docker image with that tag exists for an arbitrary or all of Certbot's DNS plugins. If it does, we exit.
3. Check if a Docker image with that tag exists for the Certbot image, if not, we send build triggers to Docker Hub to build that tag and wait for it to become available under the latest tag. (Not sure off the top of my head how to quickly check what GH tag the latest tag on Docker Hub was built with but I bet we can figure something out.)
4. Send build triggers for each of the DNS plugin images and exit.

It probably doesn't matter much, but I'd put all of the build trigger URLs in encrypted environment variables in Travis.

What do you all think?

**Comment by adferrand** [comment]
In theory, this is a pretty correct script. However in practice, your script is likely to stale for a very long time. Because in fact, DockerHub queue and speed execution on automated build is awfully slow. I am talking of potentially 30 min to wait in the queue, then 10 minutes to build (true story from my dockers).

However, Travis is very fast to build dockers (less than a minute) and to start the jobs (less than 5 minutes). Also, the docker API is very smart to avoid unnecessary local builds and network uploads/downloads: if you push something that already exist on DockerHub, nothing is uploaded, if you pull something that already exist on your local runtime, nothing is downloaded, nothing is built again.

So I would propose an idempotent workflow that ensures both the main Certbot docker and DNS plugins docker are pushed for a target tag, given a correctly checkout Certbot GIT repository.
1) Run `docker pull letsencrypt/certbot:tag`, ignoring errors, then `docker build -t letsencrypt/certbot:tag`
2) Run `docker push letsencrypt/certbot:tag` and trigger DockerHub build through API
Then, for each DNS plugin
3) Run `docker pull letsencrypt/certbot-dns-*:tag`, ignoring errors, then `docker build -t letsencrypt/certbot:tag`
4) Run `docker push letsencrypt/certbot-dns-*:tag` and trigger DockerHub build through API

The script never waits for the DockerHub automated build completness. Necessary dockers are always ensured to be pushed on DockerHub, allowing dependent dockers to build locally or on DockerHub. Builds on DockerHub are superfluous in term of build, but usefull to synchronize the relevant Dockerfile (and README for master): however, as the necessary layers will already be pushed by Travis, the job on DockerHub will run in few seconds, as nothing needs to be built again.

And of course you benefit of advanced hooks and other tests capabilities to run during Travis against the local docker images.

**Comment by bmw** [comment]
We'd need to give Travis credentials for Docker Hub to do this though and there was some concern about that above. We could get over that or alternatively run a script like the one you described locally during the release process, but the need to do that manually on a devs laptop will slow things down a fair bit.

Since we only need to update these images after a release (with the possible exception of the `rolling` tag which only exists for Certbot so we could send the build trigger on each commit and immediately exit) maybe one job waiting potentially 30 minutes once or twice a month when we do a release isn't so bad?

**Comment by bmw** [comment]
I may have just fixed this using webhooks on GitHub.

I followed the instructions at https://docs.docker.com/docker-hub/builds/classic/#adding-github-webhook-manually and set up a webhook on the Certbot repo. Merging a PR triggered a build of the rolling tag on the Certbot Docker Hub repo.

I'm a little worried about builds for tags, DNS plugins, and what we'll do if we add any more Docker images in the future, but it may be good enough for now until we figure out our path forwards for DNS plugins and Docker.

**Comment by bmw** [comment]
For some reason I don't understand, this stopped working 7 days ago :(

Docker Hub's response is still "Webhook processed" but our repository is no longer being updated.

**Comment by bmw** [comment]
For some reason Certbot is no longer configured on Docker Hub as a [classic automated build](https://docs.docker.com/docker-hub/builds/classic/) so neither webhooks or remote build triggers work anymore.

With this in mind, I propose we write a script similar to https://github.com/letsencrypt/pebble/blob/master/.ci/publish_linux.sh which we will run as part of the release process. This script should:

1. Have people login interactively their Docker Hub account.
2. Build the Certbot image using the locally tagged commit. This image should be pushed to the Docker Hub repo with the tags "latest" and one like "vx.y.z" matching the local git tag.
3. Build each image for the DNS plugin making sure it uses the recently build Certbot image as its base and pushes to the "latest" and "vx.y.z" tags in their matching Docker Hub repo.

Either included in the script or in our release script's use of it, we should make sure we have sane error handling. For instance, we should fail fast if Docker isn't installed, not exiting the entire script if someone makes a typo inputting their Docker Hub password, etc.

I think we should just do this locally (but as an automated script) as part of the release process for now and we can consider putting it in Travis with our Docker credentials in the future if we really want.

Anyone have any comments about this plan?

**Comment by ohemorange** [comment]
Currently, the release script doesn't actually push the release anywhere, it just builds the release. I assume you're proposing adding this to the release *process*, rather than the release *script*?

**Comment by bmw** [comment]
That's a good point. I suppose we probably shouldn't push anything to Docker Hub until the release has been tested...

Doing it outside of the release script makes this process slightly more annoying, but I think it's what we should do for now.

**Comment by bmw** [comment]
I edited the above proposal to reflect this.

In the future we will hopefully take the time to automate more of the release process so this can be just another automated step rather than a manual one.

**Comment by adferrand** [comment]
So with this script, the project on DockerHub will not be an automated build anymore, and will be manually updated, without Dockerfile available directly in DockerHub. @joohoi had some concerns about it. However now, the only solution will be to give credentials with write access to DockerHub over the Certbot GIT repo.

**Comment by bmw** [comment]
That's true and there is value in seeing the Dockerfile, however, Docker Hub doesn't seem to have designed this feature with security in mind (unless things have changed recently). In the past I have pushed images built locally to Certbot's Docker repository and the Dockerfile was still shown in the web UI so the user can't be sure the given Dockerfile was actually used.

I personally think we can link to the Dockerfile from the description on Docker Hub and that's good enough for now. Stopping automated builds (at least for now) also simplifies resolving https://github.com/certbot/certbot/issues/5711 because our README on GitHub is no longer automatically added to the description of the Docker image.

**Comment by joohoi** [comment]
I agree, and think it's the least bad solution that we have available right now. I'm sad to see that there really isn't a good way to actually solve this right now.

**Comment by J0WI** [comment]
Since the release process for Docker need to be changed anyway, have you considered #5540?

**Comment by bmw** [comment]
Yes. This was discussed a bit above.

We may do this in the future, but don't want to do this right now. Major aspects of Certbot are currently being redesigned such as how it is packaged and distributed, how plugins work, etc. and I personally think it makes more sense to do this work first before looking into becoming an official Docker image.

**Comment by J0WI** [comment]
The write access issue could be mitigated by moving the Docker stuff to a dedicated repo and fetching the sources from remote in the Dockerfile. So DockerHub will only have access to the Dockerfiles, but not to the main sources. I think all images in the official library use this anyway.
Multi-stage builds are not supported by the official library (https://github.com/docker-library/official-images/issues/3383), so you should avoid this if you plan to become an official image.
I think automated builds are important for transparency. This indicates that the builds are kind of reproducible. For pushed containers you have no clue who pushed what changes and you also have to trust your CI service beside Certbot and Docker. Furthermore a CI might not be aware of changes in the base images. The image should be rebuilt if Python/Alpine are updated.
I think there is also a very confusing issue on DockerHub, that the current Dockerfile and Readme will not update anymore once automated builds are disabled. So you will end up with outdated docs and sources on DckerHub.

**Comment by bmw** [comment]
That's an interesting twist to [Joona's proposal](https://github.com/certbot/certbot/issues/6630#issuecomment-454533293) of creating a readonly mirror of the Certbot repo. Thanks for the suggestion @J0WI.

My only concern with that approach initially is what triggers the Docker images to be built when we do a new release in this repository? It'd be nice if we could automate this so we don't have to do it manually.

**Comment by adferrand** [comment]
The trigger is to push a git tag to the repo.

**Comment by bmw** [comment]
How does that work if the automated build is tied to a different repository?

**Comment by adferrand** [comment]
I mean, you make operations as normal in the certbot repository, and at the end, you make a commit on the certbot-docker repo with the updated Dockerfile, and you push a git tag on it. The automated build on Dockerhub will build this tag.

**Comment by adferrand** [comment]
Note also that this certbot-docker repo needs to be in a separated github organisation, since permissions are setup at the organisation level.

**Comment by J0WI** [comment]
> Note also that this certbot-docker repo needs to be in a separated github organisation, since permissions are setup at the organisation level.

What about GitHub teams?

**Comment by adferrand** [comment]
From https://docs.docker.com/docker-hub/builds/link-source/#grant-access-to-a-github-organization, does not seem that the access can be associated to a specific team.

**Comment by adferrand** [comment]
But maybe to a dedicated user with limited access.

**Comment by joohoi** [comment]
In the light of the access privileges changes in DockerHub, the approach of separate repository (in separate organization as pointed out by @adferrand) is the best approach to handle this. After initially setting it up, we can at later point start the process of going offical DockerHub image if we wish.

DockerHub creates new builds based on tags, so it should be pretty easy to push the new tag (and update the Dockerfile if necessary) within our release process, even automatically. Though as it is being a three step process (if the Dockerfile needs to be updated, which is not very often) doesn't add too much of overhead even if it would be a manual step.

I continue to see the published `Dockerfile` as a part of an automated build a very important feature for transparency and trust.

**Comment by bmw** [comment]
Thanks for chiming in @joohoi.

I actually think we might be able to get this without creating a separate organization by creating a service account as described at https://docs.docker.com/docker-hub/builds/#service-users-for-team-autobuilds.

To test this, I created https://github.com/my-bmw-test with two repositories. Then I created a new GitHub account and only gave it admin access to https://github.com/my-bmw-test/test. The account has no special access to the organization or the other repository. Using this new GitHub account, I linked https://hub.docker.com/r/bradmw/my-certbot-test to that new GitHub account and automated builds seem to be working!

Anyone see a problem with this approach? If not, I think we should do it.

**Comment by adferrand** [comment]
Yes @bmw, this is a good approach. We still need a separate repository to host the dockerfiles, and so include that specific logic in the release process, but we can have this repo in the `certbot` organisation.

From the documentation you mentioned, it seems that this user can have only write access, not admin, to be able to trigger builds on Docker Hub.

**Comment by bmw** [comment]
> From the documentation you mentioned, it seems that this user needs only write access, not admin, to be able to trigger builds on Docker Hub.

Yeah? I didn't test it with only write permissions but in the docs I linked above it says

> This service account should have access to any repositories to be built, and must have administrative access to the source code repositories so it can manage deploy keys.

When I linked the account to Docker Hub it did add deploy keys and [this documentation](https://help.github.com/en/articles/repository-permission-levels-for-an-organization) says the GitHub account needs admin access in order to add deploy keys.

**Comment by bmw** [comment]
@adferrand is planning to write up a high level proposal for the structure of this other repo; how the files are laid out, how it gets the components of Certbot, how we update the version used, etc.

**Comment by adferrand** [comment]
So, here is my proposal.

The idea is to gather all material necessary to build the dockers in one repo, `https://github.com/certbot/certbot.git`, associate a separate GitHub user to it with write access, then use the autobuild feature in DockerHub to trigger the deployemen of all dockers (core + dns plugins) through manipulations on the GIT repository.

The advantages:
* single source of truth: GIT repository define and trigger how the Dockers are deployed
* separation of concern: a specific GIT repository is dedicated to Dockers
* security isolation: a specific user, with no access to Certbot codebase itself is used by DockerHub
* strong documentation: the Dockers repos in DockerHub are declared as with automated deployment, so in-sync description, Dockerfile, GitHub repo and tags in the DockerHub projects

A typical workflow would be this. Considering that all operations have been done to release the new `v0.35.0` version of Certbot (on certbot GIT + PyPI):
1) Checkout `master` branch of https://github.com/certbot/certbot.git
2) Make, if not already done, the appropriate corrections to the Dockerfile if necessary
3) Check that all dockers can be built correctly using `./build.sh v0.35.0`
4) Trigger the deployment by calling `./deploy.sh v0.35.0`

Remarks:
* On 2), I want to stress on the fact that the Dockerfiles should not be tied to a specific version of Certbot. The specific corrections here would be if a particular version changes the runtime (like external deps, entrypoints) and so require a modification of the Docker logic. Most of the time, the Dockerfile just do not change.
* On 4), the deployment, more precisely, would update the current version of Certbot displayed in the README.md, commit that, then tag the commit as `v0.34.0` and push it, to trigger the autobuild feature on DockerHub.

To make this work, all DockerHub projects (certbot/certbot, certbot/dns-*) will be configured to use Automated Builds, linked to the certbot docker repo https://github.com/certbot/certbot.git. One unique rule needs to be defined, depending of the nature of the Docker repo:
* for `certbot/certbot`, the rule would trigger on source `/^v[0-9.]+$/`, associating the tag `{sourceref}`, on the Dockerfile location `Dockerfile`, and Build context `/core`
* for the dns `certbot/dns-*`, the rule would trigger on source `/^v[0-9.]+$/`, associating the tag `{sourceref}`, on the Dockerfile location `Dockerfile`, and Build context `/plugin`

Using the pre/build/post hooks, DockerHub autobuild feature is able to do some preparatory steps before building and pushing the new container image:
* for `certbot/certbot`, checkout the relevant Certbot codebase from GitHub
* for the dns `certbot/dns-*`, checkout the relevant Certbot codebase from GitHub, build a local version of `certbot/certbot` (making the build independent from the effective presence of `certbot/certbot` in DockerHub), to be used as a base on the dns plugin container.

You can find a POC of this here:
* GitHub repo: https://github.com/adferrand/certbot.git
* Docker repos with tags pushed using this approach: https://hub.docker.com/r/certbot/certbot and https://hub.docker.com/certbot/dns-rfc2136

cc @bmw

**Comment by bmw** [comment]
This largely looks great! A couple questions:

> for the dns certbot/dns-*, checkout the relevant Certbot codebase from GitHub, build a local version of certbot/certbot (making the build independent from the effective presence of certbot/certbot in DockerHub)

1. What is the benefit of this approach vs. the dependency on `certbot/certbot:<tag>` you set up [here](https://github.com/adferrand/certbot-docker/blob/master/plugin/Dockerfile#L3) and [repository links](https://docs.docker.com/v17.12/docker-hub/builds/#repository-links).
2. How is your POC sharing the same README? Could we have separate READMEs for each image if we wanted?

I'd also like to hear your thoughts on the following suggestions:

1. Instead of adding pre_build hooks that tie the tag in the Docker repo to the Certbot tag, what do you think of adding a simple environment variable in the base Certbot image containing the tag/version that should be used? This variable would be updated in the `deploy.sh` script. We'd then also move the process of installing `git`, cloning the repo, installing packages, and deleting the repo and `git` dependencies into the Dockerfile. This could either be done in the Dockerfile directly or a simple script that we write that is shared between images.
2. We use capture groups in the build rules for the source tag on Docker Hub so we can make tags like `0.35.0-2` if the first build fails for some reason and Docker Hub will still automatically tag the image as `0.35.0`.
3. We do not run `./build.sh` locally before pushing the tag. If the build fails on Docker Hub, we can fix up the Docker repo and try again.

**Comment by adferrand** [comment]
I respond first to question 2.

The description is taken from the `README.md` version on the branch that has been used on the last autobuild. This means that either you need one repo for each Docker, or some weird play on specific tags for each dns plugin on one unique repo, with a specific `README.md` for each class of tags.

Both approaches seem to be unnecessarily complex, when a well written `README.md` would be suitable for all dockers.

About 1., I am not at all convinced by the linked repository feature, from the current implementation and the general approach. On the implementation, it lacks of lot of customization. What are the rules triggered in the target repo? First of them, all of them? What about rules that contain regex? From the DockerHub UI, you cannot properly trigger this kind of rules, so how we can expect particular tags matching this kind of rules to trigger a remote autobuild?

So, lack of feature, very few ways to control what is going on, which leads me to the second point.

From a general approach point of view, even if this feature was working correctly, it would mean that part of the deploy logic would be in git, and part of it would be in Docker hub. I prefer to have a single source of truth, the git repository.

**Comment by adferrand** [comment]
About suggestion 2., yes definitively I will do that!

About suggestion 3.,calling the build script is optional. Indeed, we could wait only for the status of Docker hub jobs, and redeploy if necessary. The big problem here is the "wait". On free account, only one job can be done at a time. And each job stay in the pending state at least 2 min, and I saw it go up to 10 min (with no other job in the queue). 

So, here with build.sh you have an "instant feedback" (limited by your internet connection and computational power of course) to know if the build we fail or not with the current state of the repo. I would say that is quite desirable.

**Comment by adferrand** [comment]
Finally suggestion 1. It was my initial approach. Then I realized that it was leading to a little more complexity and remove some possibility because:
- you need to install git
- you need to uninstall git and clean the repo code
- this needs to be done also on the DNS plugin, so two times overall
- what if you want to test the Docker build and behavior from a locally edited certbot repo?

That is why I decided to check out the certbot code externally, then push specific parts in the Docker during the build. Also, it is the current behavior of the Dockerfile in certbot repository.

**Comment by bmw** [comment]
> The description is taken from the README.md version on the branch that has been used on the last autobuild. This means that either you need one repo for each Docker, or some weird play on specific tags for each dns plugin on one unique repo, with a specific README.md for each class of tags.

Are you sure? Unfortunately Certbot's builds are broken, but I'm pretty sure before I edited it that it used to show the README for each plugin which matches the description at https://docs.docker.com/v17.12/docker-hub/builds/#understand-the-build-process. Maybe in addition to that, it checks the parent directories for a README as well which would explain the behavior in your POC.

We may be able to get by with a generic README, however, I want to add more information about how to use the image and I think there may be some value in removing messages about Docker images other than the one the user is currently looking at.

> About question 1., I am not at all convinced by the linked repository feature

Fine with me. I was trying to avoid reimplementing stuff, but the feature never worked right for us in the past, possibly because of things like https://github.com/docker/hub-feedback/issues/518#issuecomment-166698973 so I'm fine with sticking with the approach you have here.

> About suggestion 3.,calling the build script is optional. Indeed, we could wait only for the status of Docker hub jobs, and redeploy if necessary. The big problem here is the "wait".
> ...
> So, here with build.sh you have an "instant feedback" (limited by your internet connection and computational power of course) to know if the build we fail or not with the current state of the repo. I would say that is quite desirable.

I agree with you that there's value in this, but I was planning on making use of notifications about failed builds which eliminates waiting on either the local or remote builds. I'm not sure what your experience has been with your project, but for us, Certbot Docker builds have rarely failed (especially since we've pinned down dependencies).

> Finally suggestion 1.
> ...
> this needs to be done also on the DNS plugin, so two times overall

We could reduce some of the complexity of having to do it twice by putting it in a shared script added by the base image, but my reasons for suggesting this approach in general were:

a) It's more inline with how official images work.
b) You no longer need custom build scripts for the main Certbot image.
c) We can stop relying on `SOURCE_BRANCH` in the pre_build scripts which prevents us from doing things like `0.35.0-2`.

Point (c) is my main concern. If you change that to `DOCKER_TAG`, I'm personally fine with either approach.

**Comment by adferrand** [comment]
> Are you sure? Unfortunately Certbot's builds are broken, but I'm pretty sure before I edited it that it used to show the README for each plugin which matches the description at https://docs.docker.com/v17.12/docker-hub/builds/#understand-the-build-process. Maybe in addition to that, it checks the parent directories for a README as well which would explain the behavior in your POC.

Indeed, I checked: the README file present in the built commit relatively to the Dockerfile is taken into account, and fallback is done on the README at the root of the repository. So we can have a README for `certbot/certbot`, and one for the dns plugins. However, the dynamic modifications applied by the hooks on the README files are not taken into account. So it will be the same README for all dns plugins.

**Comment by adferrand** [comment]
> I agree with you that there's value in this, but I was planning on making use of notifications about failed builds which eliminates waiting on either the local or remote builds. I'm not sure what your experience has been with your project, but for us, Certbot Docker builds have rarely failed (especially since we've pinned down dependencies).

As I said, both approaches can be used. Building locally is very fast, and allows to quickly test some consequent modifications on the Dockerfiles. I agree it is usually not useful when the Dockerfile does not change, but it can when it does. Anyway having notifications afterward is the final judge that everything is done on DockerHub side.

**Comment by adferrand** [comment]
> We could reduce some of the complexity of having to do it twice by putting it in a shared script added by the base image, but my reasons for suggesting this approach in general were:
> 
> a) It's more inline with how official images work.
> b) You no longer need custom build scripts for the main Certbot image.
> c) We can stop relying on SOURCE_BRANCH in the pre_build scripts which prevents us from doing things like 0.35.0-2.
> 
> Point (c) is my main concern. If you change that to DOCKER_TAG, I'm personally fine with either approach.

For b) in fact you still need some customization in a way or another.

But I agree it is cleaner, and most officials images get their main piece of software from sources. I updated my repo accordingly.

**Comment by bmw** [comment]
> However, the dynamic modifications applied by the hooks on the README files are not taken into account. So it will be the same README for all dns plugins.

In the long run I'd like to the README to include a short description of each DNS plugin, a link to their readthedocs page, and instructions on how to use the image. While the dynamic Dockerfile you have now is pretty slick, I think we may want to add simple, largely duplicated Dockerfiles in the future for each DNS plugin allowing us to have separate READMEs if we wish, but I'm personally fine with the approach you have for now.

I created https://github.com/certbot/certbot-docker. Feel free to submit a PR. I think we're close enough to being on the same page that we can hammer out any details there.

**Comment by J0WI** [comment]
The image at https://hub.docker.com/r/certbot/certbot haven't been updated for 3 months now.
Could you please push an update manually?

Please consider merging #6759 and/or #7250 before doing so.

**Comment by J0WI** [comment]
No new builds have been submitted so far

**Comment by bmw** [comment]
I'm aware. Sorry for the lack of an update.

Unfortunately, it seems like the setup that I tested at https://github.com/certbot/certbot/issues/6630#issuecomment-489254660 might not actually work so we may need to do something else instead. I'm trying to verify that and then will propose another solution if that's not going to work.

**Comment by bmw** [comment]
We moved our Docker setup to https://github.com/certbot-docker/certbot-docker which is building Certbot v0.36.0 on Docker Hub right now.
